### PR TITLE
feat: built in deployment

### PR DESCRIPTION
**Pull request by 0xrusowsky**
Removes the need for users to create a specific deployment file by enhancing the contract macro (which now auto-generates the boilerplate). It achieves so with minimal changes to the user workfllow:
- target of the `deploy` binary: [before] `deploy.rs` > [now] `lib.rs`
- new feature flag: `deploy`

### CODE REVIEW

File: contract-derive/src/helpers.rs [hunk 1]
```diff
@@ -25,13 +23,33 @@ impl<'a> From<&'a TraitItemMethod> for MethodInfo<'a> {
...
-// Helper function to generate intercate impl from user-defined methods
+// Helper function to get the parameter names + types of a method
+pub fn get_arg_props<'a>(
+    skip_first_arg: bool,
+    method: &'a MethodInfo<'a>,
+) -> (Vec<Ident>, Vec<&syn::Type>) {
+    method
+        .args
+        .iter()
+        .skip(if skip_first_arg { 1 } else { 0 })
```
Comment (leonardoalt) [inline contract-derive/src/helpers.rs:40]: a bit of a pedantic comment, but not a huge fan of the if here. I think having 2 different methods would be a bit cleaner

File: contract-derive/src/helpers.rs [hunk 2]
```diff
@@ -128,3 +133,28 @@ where
...
+    constructor: Option<&ImplItemMethod>,
+) -> quote::__private::TokenStream {
+    quote! {
+        use alloc::vec::Vec;
+
+        #[no_mangle]
+        pub extern "C" fn main() -> ! {
+            // TODO: figure out constructor
+
+            let runtime: &[u8] = include_bytes!("../target/riscv64imac-unknown-none-elf/release/runtime");
```
Comment (leonardoalt) [inline contract-derive/src/helpers.rs:149]: I think this is fine for now, but we should at some point error out properly if this doesn't exist

File: contract-derive/src/helpers.rs (hunk 1 above)
Comment (0xrusowsky) [reply contract-derive/src/helpers.rs:40]: sure, np haha

would u rather:
- create 2 completely new fns (duplicating code)
- make this fn private and create `pub fn get_arg_props_all` + `pub fn get_arg_props_skip_first` effectively exposing a cleaner api, while minimizing duplicating code
- change `skip_first_arg: bool` for `args_to_skip: u8`

File: contract-derive/src/helpers.rs (hunk 2 above)
Comment (0xrusowsky) [reply contract-derive/src/helpers.rs:149]: imo the macro's default panic message i already explicit enough:
```
error: couldn't read src/../target/riscv64imac-unknown-none-elf/release/inexistent_path: No such file or directory (os error 2)
  --> src/lib.rs:42:1
   |
42 | #[contract]
   | ^^^^^^^^^^^
   |
   = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
```

thoughts?

File: contract-derive/src/helpers.rs (hunk 1 above)
Comment (0xrusowsky) [reply contract-derive/src/helpers.rs:40]: just for clarification, but we want to be able to not skip self, as it won't be an arg in the constructor

File: contract-derive/src/helpers.rs (hunk 2 above)
Comment (leonardoalt) [reply contract-derive/src/helpers.rs:149]: ah yes this looks good

File: contract-derive/src/helpers.rs (hunk 1 above)
Comment (leonardoalt) [reply contract-derive/src/helpers.rs:40]: option 2 sgtm

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**